code/modules/hydroponics/hydro_tray_process.dm

```diff
@@ -0,0 +1,340 @@
+	if(seed.spread && !closed_system && age >= seed.maturation && prob(2 * max(10,seed.potency)))
+		if((nutrilevel < 8 && waterlevel < 80) || seed.hematophage || seed.carnivorous) // Unless we're particularly vicious, let's not try to spread while our needs are met.
+			if(!(locate(/obj/effect/plantsegment) in T))
+				new /obj/effect/plantsegment(T, seed)
+				switch(seed.spread)
+					if(1)
+						msg_admin_attack("limited growth creeper vines ([seed.display_name]) have spread out of a tray. <A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[T.x];Y=[T.y];Z=[T.z]'>(JMP)</a>")
+					if(2)
+						msg_admin_attack("space vines ([seed.display_name]) have spread out of a tray. <A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[T.x];Y=[T.y];Z=[T.z]'>(JMP)</a>")
+
+	check_level_sanity()
+	update_icon()
+	return
+
+/obj/machinery/portable_atmospherics/hydroponics/proc/check_health()
+	if(health <= 0)
+		die() //ominous
+
+/obj/machinery/portable_atmospherics/hydroponics/proc/affect_growth(var/amount)
+	if(amount > 0)
+		if(age < seed.maturation)
+			age += amount
+		else if(!harvest && seed.yield != -1)
+			lastproduce -= amount
+	else
+		if(age < seed.maturation)
+			skip_aging++
+		else if(!harvest && seed.yield != -1)
+			lastproduce += amount
+
+
+/obj/machinery/portable_atmospherics/hydroponics/proc/update_name()
+	if(seed)
+		//name = "[initial(name)] ([seed.seed_name])"
+		name = "[seed.display_name]"
+	else
+		name = initial(name)
+
+	if(labeled)
+		name += " ([labeled])"
+
+//Refreshes the icon and sets the luminosity
+/obj/machinery/portable_atmospherics/hydroponics/update_icon()
+	overlays.len = 0
+
+	update_name() //fuck it i'll make it not happen constantly later
+
+	// Updates the plant overlay.
+	if(!isnull(seed))
+		if(draw_warnings && health <= (seed.endurance / 2))
+			overlays += image(seed.plant_dmi,"over_lowhealth3")
+
+		if(dead)
+			overlays += image(seed.plant_dmi,"[seed.plant_icon]-dead")
+		else if(harvest)
+			overlays += image(seed.plant_dmi,"[seed.plant_icon]-harvest")
+		else if(age < seed.maturation)
+			var/t_growthstate = max(1,round((age * seed.growth_stages) / seed.maturation))
+			overlays += image(seed.plant_dmi,"[seed.plant_icon]-grow[t_growthstate]")
+			lastproduce = age
+		else
+			overlays += image(seed.plant_dmi,"[seed.plant_icon]-grow[seed.growth_stages]")
+
+	//Draw the cover.
+	if(closed_system)
+		overlays += "hydrocover"
+
+	//Updated the various alert icons.
+	if(draw_warnings)
+		if(waterlevel <= 10)
+			overlays += "over_lowwater3"
+		if(nutrilevel <= 2)
+			overlays += "over_lownutri3"
+		if(weedlevel >= 5 || pestlevel >= 5 || toxins >= 40 || improper_heat || improper_light || improper_kpa || missing_gas)
+			overlays += "over_alert3"
+		if(harvest)
+			overlays += "over_harvest3"
+
+	// Update bioluminescence.
+	if(seed)
+		if(seed.biolum)
+			set_light(round(seed.potency/10))
+			if(seed.biolum_colour)
+				light_color = seed.biolum_colour
```

Comment:
(If I caused this when I ported lighting, sorry, I know I probably fucked up somewhere with bio luminescence)

While this would work fine (the actual value on the light source only gets updated on the next update tick), it's still not good form.